### PR TITLE
Targeting cabability for agent objectives

### PR DESCRIPTION
**Pull request by thomaswmorris**
This lets you optimize for a specific value of an objective, instead of just being able to minimize/maximize it. This is useful for e.g. getting a particular beam spot size on demand, or for "color on command" for quantum dot experiments.

### CODE REVIEW

**Comment by mrakitin** [comment]
Does this PR share commits with #49?

**Comment by mrakitin** [comment]
OK, looks like it's only 4 commits ahead of the PR #49 branch:

https://github.com/thomaswmorris/bloptools/compare/objectify...thomaswmorris:bloptools:targeting?expand=1

File: bloptools/bayesian/dofs.py [hunk 1]
```diff
@@ -0,0 +1,253 @@
...
+class DOFList(Sequence):
+    def __init__(self, dofs: list = []):
+        _validate_dofs(dofs)
+        self.dofs = dofs
+
+    def __getitem__(self, i):
+        if type(i) is int:
+            return self.dofs[i]
+        elif type(i) is str:
+            return self.dofs[self.names.index(i)]
```
Comment (mrakitin) [inline bloptools/bayesian/dofs.py:102]: The `else` case with an exception raised would help here to avoid silent `None` returns.

File: bloptools/bayesian/dofs.py [hunk 2]
```diff
@@ -0,0 +1,253 @@
...
+        return new_y
+
+
+class TimeReadback(SignalRO):
+    """
+    Returns the current timestamp.
+    """
+
+    def __init__(self, *args, **kwargs):
+        super().__init__(*args, **kwargs)
```
Comment (mrakitin) [inline bloptools/bayesian/dofs.py:236]: I think this piece does not do anything new here.

File: bloptools/bayesian/objectives.py [hunk 3]
```diff
@@ -0,0 +1,139 @@
...
+    def noise(self):
+        return self.model.likelihood.noise.item() if hasattr(self, "model") else None
+
+    @property
+    def snr(self):
+        return np.round(1 / self.model.likelihood.noise.sqrt().item(), 1) if hasattr(self, "model") else None
+
+    @property
+    def n(self):
+        return self.model.train_targets.shape[0] if hasattr(self, "model") else 0
```
Comment (mrakitin) [inline bloptools/bayesian/objectives.py:75]: Short docstrings for those properties will help.

Also, is `self.min_snr` (defaulting to `DEFAULT_MINIMUM_SNR`) used anywhere?

File: bloptools/bayesian/objectives.py [hunk 4]
```diff
@@ -0,0 +1,139 @@
...
+        return self.summary.__repr__()
+
+    # @property
+    # def descriptions(self) -> list:
+    #     return [obj.description for obj in self.objectives]
+
+    @property
+    def names(self) -> list:
+        """
+        Returns an array of the objective weights.
```
Comment (mrakitin) [inline bloptools/bayesian/objectives.py:112]:
```suggestion
        Returns an array of the objective names.
```

File: bloptools/bayesian/objectives.py [hunk 5]
```diff
@@ -0,0 +1,139 @@
...
+        """
+        return [obj.name for obj in self.objectives]
+
+    @property
+    def targets(self) -> np.array:
+        """
+        Returns an array of the objective weights.
```
Comment (mrakitin) [inline bloptools/bayesian/objectives.py:119]:
```suggestion
        Returns an array of the objective targets.
```

File: bloptools/bayesian/transforms.py [hunk 6]
```diff
@@ -0,0 +1,74 @@
...
+    r"""An affine posterior transform for scalarizing multi-output posteriors."""
+
+    scalarize: bool = True
+
+    def __init__(self, weights: Tensor, targets: Tensor) -> None:
+        r"""
+        Args:
+            weights: A one-dimensional tensor with `m` elements representing the
+                linear weights on the outputs.
+            offset: An offset to be added to posterior mean.
```
Comment (mrakitin) [inline bloptools/bayesian/transforms.py:20]: Let's fix the docstring style - looks like the class was copied over from another code base.

File: bloptools/bayesian/transforms.py [hunk 7]
```diff
@@ -0,0 +1,74 @@
+from typing import Union
+
+import botorch
+from botorch.acquisition.objective import PosteriorTransform
+from botorch.posteriors.gpytorch import GPyTorchPosterior
+from botorch.posteriors.posterior_list import PosteriorList  # pragma: no cover
```
Comment (mrakitin) [inline bloptools/bayesian/transforms.py:6]: What does `# pragma: no cover` do here?

File: bloptools/bayesian/transforms.py [hunk 8]
```diff
@@ -0,0 +1,74 @@
...
+    def mean_transform(self, mean, var):
+        for i, target in enumerate(self.targets):
+            if target == "min":
+                mean[..., i] = -mean[..., i]
+            elif target != "max":
+                mean[..., i] = -(mean[..., i] - target).abs()
+
+        return mean @ self.weights.unsqueeze(-1)
+
+    def variance_transform(self, mean, var):
```
Comment (mrakitin) [inline bloptools/bayesian/transforms.py:44]: Do we need `mean` as an input arg here? It's not used in the method...

File: bloptools/bayesian/transforms.py [hunk 9]
```diff
@@ -0,0 +1,74 @@
...
+
+    def evaluate(self, Y: Tensor) -> Tensor:
+        r"""Evaluate the transform on a set of outcomes.
+
+        Args:
+            Y: A `batch_shape x q x m`-dim tensor of outcomes.
+
+        Returns:
+            A `batch_shape x q`-dim tensor of transformed outcomes.
+        """
```
Comment (mrakitin) [inline bloptools/bayesian/transforms.py:55]: Docstring normalization is needed.

File: docs/source/tutorials/himmelblau.ipynb [hunk 10]
```diff
@@ -170,6 +170,24 @@
...
    ]
   },
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "id": "75a6a54f-0140-4bd0-8367-0dd14cdf8116",
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "{x % 2 for x in np.arange(10)}"
```
Comment (mrakitin) [inline docs/source/tutorials/himmelblau.ipynb:180]: Is this cell needed? It returns `{0, 1}`.

File: bloptools/bayesian/devices.py [hunk 11]
```diff
@@ -1,218 +1,8 @@
...
-            dof.active = True
-
-    def deactivate(self, read_only=None, active=None, tags=[]):
-        """The same as .activate(), only in reverse."""
-        for dof in self._subset_dofs(read_only, active, tags):
-            dof.active = False
+from ophyd import Signal, SignalRO  # noqa F401
 
 
 class BrownianMotion(SignalRO):
```
Comment (mrakitin) [inline bloptools/bayesian/devices.py:8]: This class is present in [`bloptools/bayesian/dofs.py`](https://github.com/NSLS-II/bloptools/pull/52/files#diff-1eee9a2e8c39b3f341e8fa40587b83e0e2f08446388f0b48155b710fe7c23706R206) along with other classes below. Did you intend to move them all together?

File: bloptools/bayesian/dofs.py [hunk 12]
```diff
@@ -0,0 +1,253 @@
...
+    #     return [device.read()[device.name]["value"] for device in self._subset_devices(read_only, active, tags)]
+
+    # def _subset_dof_names(self, read_only=None, active=None, tags=[]):
+    #     return [device.name for device in self._subset_devices(read_only, active, tags)]
+
+    # def _subset_dof_limits(self, read_only=None, active=None, tags=[]):
+    #     dofs_subset = self._subset_dofs(read_only, active, tags)
+    #     if len(dofs_subset) > 0:
+    #         return torch.tensor([dof["limits"] for dof in dofs_subset], dtype=torch.float64).T
+    #     return torch.empty((2, 0))
```
Comment (mrakitin) [inline bloptools/bayesian/dofs.py:195]: Probably those commented lines can be removed.

**Review by mrakitin — COMMENTED** [review]
The conversion of the DOFs and objectives to Dataclasses looks great! I spotted a few non-critical issues which would be nice to address before merging.

File: bloptools/bayesian/objectives.py (hunk 3 above)
Comment (thomaswmorris) [reply bloptools/bayesian/objectives.py:75]: Yes, it's used in the agent

File: bloptools/bayesian/transforms.py (hunk 8 above)
Comment (thomaswmorris) [reply bloptools/bayesian/transforms.py:44]: Yes, BoTorch requires mean and var as inputs

File: docs/source/tutorials/himmelblau.ipynb (hunk 10 above)
Comment (thomaswmorris) [reply docs/source/tutorials/himmelblau.ipynb:180]: no, I'll fix the notebooks
